lib/rust/parser/src/format.rs

```diff
@@ -0,0 +1,529 @@
+        self.finish()
+    }
+}
+
+impl SerializeSeq for ObjectSerializer<'_> {
+    type Ok = Ok;
+    type Error = Error;
+
+    fn serialize_element<T>(&mut self, value: &T) -> Result<()>
+    where T: ?Sized + Serialize {
+        value.serialize(&mut *self.serializer)
+    }
+
+    fn end(self) -> Result<()> {
+        self.finish()
+    }
+}
+
+impl ser::SerializeTupleVariant for ObjectSerializer<'_> {
+    type Ok = Ok;
+    type Error = Error;
+
+    fn serialize_field<T>(&mut self, value: &T) -> Result<()>
+    where T: ?Sized + Serialize {
+        value.serialize(&mut *self.serializer)
+    }
+
+    fn end(self) -> Result<()> {
+        self.finish()
+    }
+}
+
+impl ser::SerializeMap for ObjectSerializer<'_> {
+    type Ok = Ok;
+    type Error = Error;
+
+    fn serialize_key<T>(&mut self, key: &T) -> Result<()>
+    where T: ?Sized + Serialize {
+        key.serialize(&mut *self.serializer)
+    }
+
+    fn serialize_value<T>(&mut self, value: &T) -> Result<()>
+    where T: ?Sized + Serialize {
+        value.serialize(&mut *self.serializer)
+    }
+
+    fn end(self) -> Result<()> {
+        self.finish()
+    }
+}
+
+
+
+// ====================
+// === Result Types ===
+// ====================
+
+type Ok = ();
+
+/// Describes a serialization failure.
+#[derive(Debug)]
+pub enum Error {
+    /// Indicates that the nested object depth of the input exceeded [`RECURSION_LIMIT`], and
+    /// serialization was aborted to prevent a stack overflow. This is not expected to occur for
+    /// "reasonable" syntax trees.
+    RecursionLimitExceeded,
+    /// A serialization failure described by a message.
+    Custom(String),
+}
+
+impl ser::StdError for Error {}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
+impl ser::Error for Error {
+    fn custom<T>(msg: T) -> Self
+    where T: Display {
+        Self::Custom(msg.to_string())
+    }
+}
+
+/// The result of a serialization attempt.
+pub type Result<T> = std::result::Result<T, Error>;
+
+
+// =============
+// === Tests ===
+// =============
+
+#[cfg(test)]
+mod test {
+    use serde::Serialize;
+
+    #[test]
+    fn test_infinite_recursion() {
+        use std::cell::RefCell;
+        use std::rc::Rc;
+        /// A serializable object containing a reference to itself.
+        #[derive(Serialize)]
+        struct Cyclic {
+            this: RefCell<Option<Rc<Cyclic>>>,
+        }
+        impl Cyclic {
+            fn new() -> Rc<Self> {
+                let cyclic = Rc::new(Cyclic { this: RefCell::new(None) });
+                *cyclic.this.borrow_mut() = Some(Rc::clone(&cyclic));
+                cyclic
+            }
+        }
+        // Note that if recursion is not adequately limited the expected failure mode is aborting
+        // due to stack overflow. We are just checking `is_err` here for good measure.
+        assert!(super::serialize(Cyclic::new()).is_err());
+    }
```

Comment:
Add a test case for serializing some simple structure, and also simple but with boxed fields structure (checking the generated binary content).